### PR TITLE
Use ArmeriaServerGracefulShutdownLifecycle to start Server with Spring integration

### LINKED ISSUE #3759: Use `ArmeriaServerGracefulShutdownLifecycle` to start `Server` with Spring integration
Armeria `Server` is started before Spring Application contexts are fully initialized when using the Spring integration module.
https://github.com/line/armeria/blob/be2bf5176dc20cc868011e9239510f2feac87328/spring/boot2-autoconfigure/src/main/java/com/linecorp/armeria/spring/AbstractArmeriaAutoConfiguration.java#L104-L110
If a `Server` early binds to a port before the application is ready, many pending requests are queued and it will cause GC pressure.
I think we can use `ArmeriaServerGracefulShutdownLifecycle` added in #3226 to start up a `Server` leveraging Spring beans' lifecycle. 
https://github.com/line/armeria/blob/f3cc329530a83ec47ed08c8e8be1443437e55de7/spring/boot2-autoconfigure/src/main/java/com/linecorp/armeria/spring/ArmeriaServerGracefulShutdownLifecycle.java#L36-L42

### PR DESCRIPTION
**Pull request by heowc**
Motivation:

When integrating the armeria and the spring, many requests are pending, resulting in GC pressure.

Modifications:

- Change the armeria start point to `ArmeriaServerGracefulShutdownLifecycle`.

Result:

- Closes #3759

### CODE REVIEW

File: spring/boot2-autoconfigure/src/main/java/com/linecorp/armeria/spring/ArmeriaServerGracefulShutdownLifecycle.java [hunk 1]
```diff
@@ -27,7 +29,11 @@
  * So Armeria will shutdown before other web servers and beans in the context.
  */
 final class ArmeriaServerGracefulShutdownLifecycle implements SmartLifecycle {
+
+    private final Logger logger = LoggerFactory.getLogger(getClass());
```
Comment (ikhoon) [inline spring/boot2-autoconfigure/src/main/java/com/linecorp/armeria/spring/ArmeriaServerGracefulShutdownLifecycle.java:33]:
```suggestion
    private static final Logger logger = LoggerFactory.getLogger(ArmeriaServerGracefulShutdownLifecycle.class);
```

File: spring/boot2-autoconfigure/src/main/java/com/linecorp/armeria/spring/ArmeriaServerGracefulShutdownLifecycle.java [hunk 2]
```diff
@@ -39,6 +45,14 @@
      */
     @Override
     public void start() {
```
Comment (ikhoon) [inline spring/boot2-autoconfigure/src/main/java/com/linecorp/armeria/spring/ArmeriaServerGracefulShutdownLifecycle.java:47]: Could you update the Javadoc which is invalid now?

File: spring/boot2-autoconfigure/src/main/java/com/linecorp/armeria/spring/ArmeriaServerGracefulShutdownLifecycle.java [hunk 3]
```diff
@@ -82,6 +97,6 @@ public boolean isRunning() {
      */
     @Override
     public boolean isAutoStartup() {
```
Comment (ikhoon) [inline spring/boot2-autoconfigure/src/main/java/com/linecorp/armeria/spring/ArmeriaServerGracefulShutdownLifecycle.java:99]: Please update the Javadoc. 🙇‍♂️

File: spring/boot2-autoconfigure/src/main/java/com/linecorp/armeria/spring/ArmeriaServerGracefulShutdownLifecycle.java [hunk 4]
```diff
@@ -27,18 +29,29 @@
...
-     * Currently AbstractArmeriaAutoConfiguration help starting the server.
      */
     @Override
     public void start() {
+        running = true;
+        server.start().handle((result, t) -> {
+            if (t != null) {
+                throw new IllegalStateException("Armeria server failed to start", t);
+            }
+            return result;
```
Comment (minwoox) [inline spring/boot2-autoconfigure/src/main/java/com/linecorp/armeria/spring/ArmeriaServerGracefulShutdownLifecycle.java:52]: nit:
```suggestion
            return null;
```

File: spring/boot2-autoconfigure/src/main/java/com/linecorp/armeria/spring/ArmeriaServerGracefulShutdownLifecycle.java [hunk 5]
```diff
@@ -27,18 +29,29 @@
...
         this.server = requireNonNull(server, "server");
     }
 
     /**
      * Start this component.
-     * Currently AbstractArmeriaAutoConfiguration help starting the server.
      */
     @Override
     public void start() {
+        running = true;
```
Comment (minwoox) [inline spring/boot2-autoconfigure/src/main/java/com/linecorp/armeria/spring/ArmeriaServerGracefulShutdownLifecycle.java:47]: Shouldn't we set this at line 51?
```
server.start().handle((result, t) -> {
    if (t != null) {
        throw new IllegalStateException("Armeria server failed to start", t);
    }
    running = true;
    ...
```

**Review by minwoox — APPROVED** [review]
Left some minor comments. Thanks!

File: spring/boot2-autoconfigure/src/main/java/com/linecorp/armeria/spring/ArmeriaServerGracefulShutdownLifecycle.java (hunk 5 above)
Comment (heowc) [reply spring/boot2-autoconfigure/src/main/java/com/linecorp/armeria/spring/ArmeriaServerGracefulShutdownLifecycle.java:47]: That seems more correct. 🤣

**Review by trustin — APPROVED** [review]
Thanks, @heowc!

**Pull request merged**